### ISSUE TITLE
Cannot remove public/assets/ from linked_dirs

### ISSUE THREAD

**Issue by soundasleep** [capistrano/rails#218]
I'm using Capistrano to deploy a site that includes JS/CSS built with Gulp, outside of the Rails assets pipeline.

If I build with gulp to `public/assets` (which makes sense to me), then since `public/assets` [is in linked_dirs by default](https://github.com/capistrano/rails/issues/101), if I'm not careful I can end up with a never-ending collection of files in that folder.

I tried removing `public/assets` from `linked_dirs` in `config/deploy.rb` with:

```ruby
remove :linked_dirs, "public/assets"
```

but this didn't work; after deploy, the folder is still symlinked. So I have two issues:

1. How does one remove `public/assets` from `linked_dirs`?

2. Why is `public/assets` in `linked_dirs` anyway? Going through git blames I couldn't find a reason why.

**Comment by will-in-wi** [comment]
A workaround to do this would be to create a new task to remove `public/assets` from `:linked_dirs` after `deploy:set_linked_dirs`.

Something like:

```ruby
namespace :deploy do
  task :remove_linked_dirs do
    remove :linked_dirs, "public/assets"
  end
end
after 'deploy:set_linked_dirs', 'deploy:remove_linked_dirs'
```

**Comment by will-in-wi** [comment]
I don't see a reason in the blames either. However, I can speculate that if you do not share the assets directory, if a cached page load tries to fetch an old asset, you don't necessarily want that to suddenly fail.

**Comment by will-in-wi** [comment]
I'd probably suggest in your situation to stick the files somewhere other than `public/assets`. If you aren't using the asset pipeline, it probably makes more sense to just use a new directory.

**Comment by mattbrictson** [comment]
> I can speculate that if you do not share the assets directory, if a cached page load tries to fetch an old asset, you don't necessarily want that to suddenly fail.

Yep, you pretty much nailed it. From the [official Rails documentation](http://guides.rubyonrails.org/asset_pipeline.html#precompiling-assets):

> It is important that this folder is shared between deployments so that remotely cached pages referencing the old compiled assets still work for the life of the cached page.

**Comment by aesyondu** [comment]
> A workaround to do this would be to create a new task to remove `public/assets` from `:linked_dirs` after `deploy:set_linked_dirs`.
> 
> Something like:
> 
> ```ruby
> namespace :deploy do
>   task :remove_linked_dirs do
>     remove :linked_dirs, "public/assets"
>   end
> end
> after 'deploy:set_linked_dirs', 'deploy:remove_linked_dirs'
> ```

Thanks. Literally copy pasted this and it worked.

**Comment by aesyondu** [comment]
> > It is important that this folder is shared between deployments so that remotely cached pages referencing the old compiled assets still work for the life of the cached page.

Can someone expand on this further? Why would it be a good idea for cached pages to to be able to access old compiled assets? Wouldn't you want your users to always receive the latest assets, which is the entire point of cache busting?

**Comment by will-in-wi** [comment]
> Can someone expand on this further? Why would it be a good idea for cached pages to to be able to access old compiled assets? Wouldn't you want your users to always receive the latest assets, which is the entire point of cache busting?

Good question!

In short, since HTML, CSS, JS, et cetera are all sent down separately, there are often cases where you are loading a cached HTML file but do not also have the other assets cached. If those URLs are no longer valid, then you will have a broken, unstyled page.

This can happen in a number of different ways. One situation at a company I used to work for involved serving the website from a server cluster behind Akamai. If old assets had just disappeared on deploy, Akamai could have a recent, pre-deploy cached copy of the HTML, but have just after the deploy checked the origin for the CSS, seen the 404, and dumped its cached copy. This would result in up to ten minutes (in our case) of unstyled websites displaying until Akamai caught up (or someone noticed and manually dumped the cache). I used Capistrano there, and on a new website, we actually accidentally encountered this in production.
